File: app/src/main/java/com/kirchhoff/example/githubclient/repository/GitHubRepository.java
package com.kirchhoff.example.githubclient.repository;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.kirchhoff.example.githubclient.Injection;
import com.kirchhoff.example.githubclient.api.ApiFactory;
import com.kirchhoff.example.githubclient.model.Authorization;
import com.kirchhoff.example.githubclient.model.CommitResponse;
import com.kirchhoff.example.githubclient.model.Repository;
import com.kirchhoff.example.githubclient.utils.AuthorizationUtils;

import java.util.List;

import rx.Observable;

public class GitHubRepository implements GitHubDataSource {

    @Nullable
    private static GitHubRepository INSTANCE = null;

    private GitHubRepository() {
    }

    public static GitHubRepository getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new GitHubRepository();
        }

        return INSTANCE;
    }

    @NonNull
    @Override
    public Observable<Authorization> auth(@NonNull String login, @NonNull String password) {

        String authorizationString = AuthorizationUtils.createAuthorizationString(login, password);

        return ApiFactory.getGitHubService().authorize(authorizationString,
                AuthorizationUtils.createAuthorizationParam()).doOnNext(authorization -> ApiFactory.recreate());
    }

    @NonNull
    @Override
    public Observable<List<Repository>> repositories() {
        return ApiFactory.getGitHubService().repositories();
    }

    @NonNull
    @Override
    public Observable<List<CommitResponse>> getCommits(@NonNull String repos) {
        return ApiFactory.getGitHubService()
                .commits(Injection.provideKeyValueStorage().getUserName(), repos);
    }

    @Override
    public void logout() {
        ApiFactory.recreate();
    }
}

File: app/src/main/java/com/kirchhoff/example/githubclient/repository/GitHubRepository.kt
package com.kirchhoff.example.githubclient.repository

import com.kirchhoff.example.githubclient.Injection
import com.kirchhoff.example.githubclient.api.ApiFactory
import com.kirchhoff.example.githubclient.model.Authorization
import com.kirchhoff.example.githubclient.model.CommitResponse
import com.kirchhoff.example.githubclient.model.Repository
import com.kirchhoff.example.githubclient.utils.AuthorizationUtils
import rx.Observable

/**
 * @author Kirchhoff-
 */
object GitHubRepository : GitHubDataSource {

    override fun auth(login: String, password: String): Observable<Authorization> {
        val authorizationString = AuthorizationUtils.createAuthorizationString(login, password)

        return ApiFactory.getGitHubService().authorize(authorizationString,
                AuthorizationUtils.createAuthorizationParam()).doOnNext { ApiFactory.recreate() }
    }

    override fun repositories(): Observable<List<Repository>> =
            ApiFactory.getGitHubService().repositories()

    override fun getCommits(repos: String): Observable<List<CommitResponse>> =
            ApiFactory.getGitHubService().commits(Injection.provideKeyValueStorage().userName, repos)

    override fun logout() {
        ApiFactory.recreate()
    }

}